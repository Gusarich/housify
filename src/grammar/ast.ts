export type AstModule = {
    kind: 'module';
    items: AstModuleItem[];
    id: number;
};

export type AstModuleItem = AstHouse;

export type AstHouse = {
    kind: 'house';
    name: AstId;
    items: AstHouseItem[];
    id: number;
};

export type AstHouseItem = AstGlobalStat | AstPlayerStat | AstHandler;

export type AstGlobalStat = {
    kind: 'globalStat';
    name: AstId;
    type: AstId;
    defaultValue?: AstExpression;
    id: number;
};

export type AstPlayerStat = {
    kind: 'playerStat';
    name: AstId;
    type: AstId;
    defaultValue?: AstExpression;
    id: number;
};

export type AstHandler = {
    kind: 'handler';
    event: AstId;
    statements: AstStatement[];
    id: number;
};

export type AstStatement =
    | AstStatementAssign
    | AstStatementLet
    | AstStatementExpression
    | AstStatementIf;

export type AstStatementAssign = {
    kind: 'statementAssign';
    lvalue: AstExpression;
    value: AstExpression;
    id: number;
};

export type AstStatementLet = {
    kind: 'statementLet';
    name: AstId;
    type: AstId;
    value: AstExpression;
    id: number;
};

export type AstStatementExpression = {
    kind: 'statementExpression';
    expression: AstExpression;
    id: number;
};

export type AstStatementIf = {
    kind: 'statementIf';
    condition: AstExpression;
    then: AstStatement[];
    else?: AstStatement[];
    id: number;
};

export type AstExpression =
    | AstExpressionBinary
    | AstExpressionUnary
    | AstIntegerLiteral
    | AstExpressionId;

export type AstBinaryOp =
    | '+'
    | '-'
    | '*'
    | '/'
    | '=='
    | '!='
    | '<'
    | '<='
    | '>'
    | '>='
    | '&&'
    | '||';

export type AstExpressionBinary = {
    kind: 'expressionBinary';
    op: AstBinaryOp;
    left: AstExpression;
    right: AstExpression;
    id: number;
};

export type AstUnaryOp = '+' | '-' | '!';

export type AstExpressionUnary = {
    kind: 'expressionUnary';
    op: AstUnaryOp;
    operand: AstExpression;
    id: number;
};

export type AstIntegerLiteral = {
    kind: 'integerLiteral';
    value: string;
    id: number;
};

export type AstExpressionId = {
    kind: 'expressionId';
    name: AstId;
    id: number;
};

export type AstId = {
    kind: 'id';
    name: string;
    id: number;
};

export type AstNode =
    | AstModule
    | AstModuleItem
    | AstHouseItem
    | AstStatement
    | AstExpression
    | AstId;

let nextId = 1;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;

export function createAstNode(src: DistributiveOmit<AstNode, 'id'>): AstNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function cloneAstNode<T extends AstNode>(src: T): T {
    return { ...src, id: nextId++ };
}

export function resetNodeId() {
    nextId = 1;
}
