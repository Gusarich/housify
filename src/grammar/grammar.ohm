Housify {
    Module = ModuleItem*

    ModuleItem = House
    
    House = house id "{" HouseItem* "}"

    HouseItem = GlobalStat | PlayerStat | Handler

    GlobalStat = global id ":" Type ("=" Expression)? ";"
    PlayerStat = player id ":" Type ("=" Expression)? ";"

    Handler = handle id "{" Statement* "}"

    Statement = StatementAssign
              | StatementLet
              | StatementExpression
              | StatementIf
    
    StatementAssign = Expression "=" Expression ";"

    StatementLet = let id ":" Type "=" Expression ";"

    StatementExpression = Expression ";"

    StatementIf = if Expression "{" Statement* "}" ~else --noElse
                | if Expression "{" Statement* "}" else "{" Statement* "}" --withElse
                | if Expression "{" Statement* "}" else StatementIf --withElseIf
    
    Expression = ExpressionBinary
               | ExpressionUnary
               | ExpressionLiteral
               | ExpressionId
               | ExpressionParens
    
    ExpressionBinary = Expression binaryOp Expression
    ExpressionUnary = unaryOp Expression
    ExpressionLiteral = integerLiteral
    ExpressionId = id
    ExpressionParens = "(" Expression ")"

    binaryOp = "+" | "-" | "*" | "/" | "==" | "!=" | "<" | "<=" | ">" | ">=" | "&&" | "||"
    unaryOp = "+" | "-" | "!"

    Type = id --regular

    integerLiteral = digit+

    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC
    letterComment = letterAsciiLC | letterAsciiUC | digit | "_"

    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~keyword #idStart #(idPart*)

    house = "house" ~idPart
    global = "global" ~idPart
    player = "player" ~idPart
    handle = "handle" ~idPart
    let = "let" ~idPart
    if = "if" ~idPart
    else = "else" ~idPart

    keyword = house | global | player | handle | let | if | else
}