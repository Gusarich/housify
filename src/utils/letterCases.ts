export function snakeToCamel(s: string) {
    return s.replace(/(_\w)/g, (m) => m[1]!.toUpperCase());
}

export function camelToSnake(s: string) {
    return s.replace(/[A-Z]/g, (m) => '_' + m.toLowerCase()).slice(1);
}
