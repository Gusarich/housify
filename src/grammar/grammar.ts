import { IterationNode } from 'ohm-js';
import {
    AstBinaryOp,
    AstModule,
    AstNode,
    AstUnaryOp,
    createAstNode,
} from './ast';
import grammar from './grammar.ohm-bundle';

function unwrapOptNode(node: IterationNode): IterationNode | undefined {
    return node.children.length === 0 ? undefined : node.children[0];
}

const semantics = grammar.createSemantics();

semantics.addOperation<AstNode>('astOfModule', {
    Module(items) {
        return createAstNode({
            kind: 'module',
            items: items.children.map((item) => item.astOfModuleItem()),
        });
    },
});

semantics.addOperation<AstNode>('astOfModuleItem', {
    ModuleItem(arg0) {
        return arg0.astOfModuleItem();
    },
    House(_arg0, id, _arg2, itemsTop, items, _arg4) {
        return createAstNode({
            kind: 'house',
            name: id.astOfExpression(),
            items: itemsTop.children
                .map((item) => item.astOfHouseItem())
                .concat(items.children.map((item) => item.astOfHouseItem())),
        });
    },
});

semantics.addOperation<AstNode>('astOfHouseItem', {
    HouseItem(arg0) {
        return arg0.astOfHouseItem();
    },
    HouseItemTop(arg0) {
        return arg0.astOfHouseItem();
    },
    GlobalStat(_arg0, name, _arg2, type, _arg4, defaultValueOpt, _arg6) {
        return createAstNode({
            kind: 'globalStat',
            name: name.astOfExpression(),
            type: type.astOfType(),
            defaultValue: unwrapOptNode(defaultValueOpt)?.astOfExpression(),
        });
    },
    PlayerStat(_arg0, name, _arg2, type, _arg4, defaultValueOpt, _arg6) {
        return createAstNode({
            kind: 'playerStat',
            name: name.astOfExpression(),
            type: type.astOfType(),
            defaultValue: unwrapOptNode(defaultValueOpt)?.astOfExpression(),
        });
    },
    Handler(_arg0, _arg1, eventParameter, _arg3, _arg4, statements, _arg6) {
        return createAstNode({
            kind: 'handler',
            event: eventParameter.astOfParameter(),
            statements: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
        });
    },
});

semantics.addOperation<AstNode>('astOfParameter', {
    Parameter(arg0, _arg1, arg2) {
        return createAstNode({
            kind: 'parameter',
            name: arg0.astOfExpression(),
            type: arg2.astOfType(),
        });
    },
});

semantics.addOperation<AstNode>('astOfType', {
    Type(arg0) {
        return arg0.astOfType();
    },
    Type_regular(arg0) {
        return arg0.astOfExpression();
    },
});

semantics.addOperation<AstNode>('astOfStatement', {
    Statement(arg0) {
        return arg0.astOfStatement();
    },
    StatementAssign(lvalue, _arg1, expression, _arg3) {
        return createAstNode({
            kind: 'statementAssign',
            lvalue: lvalue.astOfExpression(),
            value: expression.astOfExpression(),
        });
    },
    StatementLet(_arg0, name, _arg2, type, _arg4, expression, _arg6) {
        return createAstNode({
            kind: 'statementLet',
            name: name.astOfExpression(),
            type: type.astOfType(),
            value: expression.astOfExpression(),
        });
    },
    StatementExpression(expression, _arg1) {
        return createAstNode({
            kind: 'statementExpression',
            expression: expression.astOfExpression(),
        });
    },
    StatementIf(arg0) {
        return arg0.astOfStatement();
    },
    StatementIf_noElse(_arg0, condition, _arg2, statements, _arg4) {
        return createAstNode({
            kind: 'statementIf',
            condition: condition.astOfExpression(),
            then: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
        });
    },
    StatementIf_withElse(
        _arg0,
        condition,
        _arg2,
        statements,
        _arg4,
        _arg5,
        _arg6,
        elseStatements,
        _arg8,
    ) {
        return createAstNode({
            kind: 'statementIf',
            condition: condition.astOfExpression(),
            then: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
            else: elseStatements.children.map((statement) =>
                statement.astOfStatement(),
            ),
        });
    },
    StatementIf_withElseIf(
        _arg0,
        condition,
        _arg2,
        statements,
        _arg4,
        _arg5,
        elseif,
    ) {
        return createAstNode({
            kind: 'statementIf',
            condition: condition.astOfExpression(),
            then: statements.children.map((statement) =>
                statement.astOfStatement(),
            ),
            else: [elseif.astOfStatement()],
        });
    },
});

semantics.addOperation<AstNode>('astOfExpression', {
    Expression(arg0) {
        return arg0.astOfExpression();
    },
    ExpressionBinary(left, op, right) {
        return createAstNode({
            kind: 'expressionBinary',
            op: op.astOfBinaryOp(),
            left: left.astOfExpression(),
            right: right.astOfExpression(),
        });
    },
    ExpressionUnary(op, operand) {
        return createAstNode({
            kind: 'expressionUnary',
            op: op.astOfUnaryOp(),
            operand: operand.astOfExpression(),
        });
    },
    ExpressionLiteral(arg0) {
        return arg0.astOfExpression();
    },
    ExpressionId(arg0) {
        return createAstNode({
            kind: 'expressionId',
            name: arg0.astOfExpression(),
        });
    },
    ExpressionField(struct, _arg1, field) {
        return createAstNode({
            kind: 'expressionField',
            struct: struct.astOfExpression(),
            field: field.astOfExpression(),
        });
    },
    id(_arg0) {
        return createAstNode({
            kind: 'id',
            name: this.sourceString,
        });
    },
    integerLiteral(_arg0) {
        return createAstNode({
            kind: 'integerLiteral',
            value: this.sourceString,
        });
    },
    booleanLiteral(_arg0) {
        return createAstNode({
            kind: 'booleanLiteral',
            value: this.sourceString === 'true',
        });
    },
    ExpressionParens(_arg0, expression, _arg2) {
        return expression.astOfExpression();
    },
});

semantics.addOperation<AstBinaryOp>('astOfBinaryOp', {
    binaryOp(op) {
        switch (op.sourceString) {
            case '+':
            case '-':
            case '*':
            case '/':
            case '==':
            case '!=':
            case '<':
            case '<=':
            case '>':
            case '>=':
            case '&&':
            case '||':
                return op.sourceString as AstBinaryOp;
            default:
                throw new Error(
                    `Unexpected binary operator: '${op.sourceString}'`,
                );
        }
    },
});

semantics.addOperation<AstUnaryOp>('astOfUnaryOp', {
    unaryOp(op) {
        switch (op.sourceString) {
            case '+':
            case '-':
            case '!':
                return op.sourceString as AstUnaryOp;
            default:
                throw new Error(
                    `Unexpected unary operator: '${op.sourceString}'`,
                );
        }
    },
});

export function parse(src: string): AstModule {
    const matchResult = grammar.match(src);
    if (matchResult.failed()) {
        throw Error(matchResult.message);
    }
    return semantics(matchResult).astOfModule();
}
