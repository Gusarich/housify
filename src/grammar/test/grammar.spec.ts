import { parse } from '../grammar';
import { loadCases } from '../../utils/loadCases';
import { resetNodeId } from '../ast';

describe('Grammar', () => {
    beforeEach(() => {
        resetNodeId();
    });

    for (const r of loadCases(__dirname + '/cases/')) {
        it('should parse ' + r.name, () => {
            expect(parse(r.code)).toMatchSnapshot();
        });
    }

    for (const r of loadCases(__dirname + '/cases-failed/')) {
        it('should fail ' + r.name, () => {
            expect(() => parse(r.code)).toThrowErrorMatchingSnapshot();
        });
    }
});
