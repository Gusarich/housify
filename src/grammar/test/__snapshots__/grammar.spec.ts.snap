// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grammar should fail 1 1`] = `
"Line 5, col 15:
  4 |     player xp: int = 123;
> 5 |     player yp int;
                    ^
  6 | 
Expected ":""
`;

exports[`Grammar should fail 2 1`] = `
"Line 3, col 13:
  2 |     handle(event: BlockBreak) {
> 3 |         x = ;
                  ^
  4 |     }
Expected "(", "player", "global", "_", "A".."Z", "a".."z", "false", "true", a digit, "!", or "-""
`;

exports[`Grammar should fail 3 1`] = `
"Line 3, col 22:
  2 |     handle(event: BlockBreak) {
> 3 |         let x: int = ;
                           ^
  4 |     }
Expected "(", "player", "global", "_", "A".."Z", "a".."z", "false", "true", a digit, "!", or "-""
`;

exports[`Grammar should fail 4 1`] = `
"Line 4, col 5:
  3 |         5
> 4 |     }
          ^
  5 | }
Expected ";", "=", "||", "&&", ">=", ">", "<=", "<", "!=", "==", "/", "*", "-", "+", or ".""
`;

exports[`Grammar should fail 5 1`] = `
"Line 6, col 5:
  5 |         } else
> 6 |     }
          ^
  7 | }
Expected "if" or "{""
`;

exports[`Grammar should fail 6 1`] = `
"Line 3, col 17:
  2 |     handle(event: BlockBreak) {
> 3 |         x = 5 + ;
                      ^
  4 |     }
Expected "(", "player", "global", "_", "A".."Z", "a".."z", "false", "true", a digit, "!", or "-""
`;

exports[`Grammar should fail 7 1`] = `
"Line 3, col 14:
  2 |     handle(event: BlockBreak) {
> 3 |         x = -;
                   ^
  4 |     }
Expected "(", "player", "global", "_", "A".."Z", "a".."z", "false", "true", a digit, "!", or "-""
`;

exports[`Grammar should fail 8 1`] = `
"Line 3, col 15:
  2 |     handle(event: BlockBreak) {
> 3 |         x = 42a;
                    ^
  4 |     }
Expected ";""
`;

exports[`Grammar should fail 9 1`] = `
"Line 4, col 5:
  3 |         myVar
> 4 |     }
          ^
  5 | }
Expected ";", "=", "||", "&&", ">=", ">", "<=", "<", "!=", "==", "/", "*", "-", "+", or ".""
`;

exports[`Grammar should fail 10 1`] = `
"Line 3, col 19:
  2 |     handle(event: BlockBreak) {
> 3 |         (x + y) * ;
                        ^
  4 |     }
Expected "(", "player", "global", "_", "A".."Z", "a".."z", "false", "true", a digit, "!", or "-""
`;

exports[`Grammar should fail 11 1`] = `
"Line 10, col 5:
   9 | 
> 10 |     player yp: int;
           ^
  11 | }
Expected "}" or "handle""
`;

exports[`Grammar should parse 1 1`] = `
{
  "id": 21,
  "items": [
    {
      "id": 20,
      "items": [
        {
          "defaultValue": {
            "id": 4,
            "kind": "integerLiteral",
            "value": "10",
          },
          "id": 5,
          "kind": "globalStat",
          "name": {
            "id": 2,
            "kind": "id",
            "name": "xg",
          },
          "type": {
            "id": 3,
            "kind": "id",
            "name": "int",
          },
        },
        {
          "defaultValue": undefined,
          "id": 8,
          "kind": "globalStat",
          "name": {
            "id": 6,
            "kind": "id",
            "name": "yg",
          },
          "type": {
            "id": 7,
            "kind": "id",
            "name": "int",
          },
        },
        {
          "defaultValue": {
            "id": 11,
            "kind": "integerLiteral",
            "value": "123",
          },
          "id": 12,
          "kind": "playerStat",
          "name": {
            "id": 9,
            "kind": "id",
            "name": "xp",
          },
          "type": {
            "id": 10,
            "kind": "id",
            "name": "int",
          },
        },
        {
          "defaultValue": undefined,
          "id": 15,
          "kind": "playerStat",
          "name": {
            "id": 13,
            "kind": "id",
            "name": "yp",
          },
          "type": {
            "id": 14,
            "kind": "id",
            "name": "int",
          },
        },
        {
          "event": {
            "id": 18,
            "kind": "parameter",
            "name": {
              "id": 16,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 17,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 19,
          "kind": "handler",
          "statements": [],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 2 1`] = `
{
  "id": 11,
  "items": [
    {
      "id": 10,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 9,
          "kind": "handler",
          "statements": [
            {
              "id": 8,
              "kind": "statementAssign",
              "lvalue": {
                "id": 6,
                "kind": "expressionId",
                "name": {
                  "id": 5,
                  "kind": "id",
                  "name": "x",
                },
              },
              "value": {
                "id": 7,
                "kind": "integerLiteral",
                "value": "5",
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 3 1`] = `
{
  "id": 11,
  "items": [
    {
      "id": 10,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 9,
          "kind": "handler",
          "statements": [
            {
              "id": 8,
              "kind": "statementLet",
              "name": {
                "id": 5,
                "kind": "id",
                "name": "x",
              },
              "type": {
                "id": 6,
                "kind": "id",
                "name": "int",
              },
              "value": {
                "id": 7,
                "kind": "integerLiteral",
                "value": "5",
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 4 1`] = `
{
  "id": 9,
  "items": [
    {
      "id": 8,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 7,
          "kind": "handler",
          "statements": [
            {
              "expression": {
                "id": 5,
                "kind": "integerLiteral",
                "value": "5",
              },
              "id": 6,
              "kind": "statementExpression",
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 5 1`] = `
{
  "id": 16,
  "items": [
    {
      "id": 15,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 14,
          "kind": "handler",
          "statements": [
            {
              "condition": {
                "id": 8,
                "kind": "expressionBinary",
                "left": {
                  "id": 6,
                  "kind": "expressionId",
                  "name": {
                    "id": 5,
                    "kind": "id",
                    "name": "x",
                  },
                },
                "op": "==",
                "right": {
                  "id": 7,
                  "kind": "integerLiteral",
                  "value": "5",
                },
              },
              "id": 13,
              "kind": "statementIf",
              "then": [
                {
                  "id": 12,
                  "kind": "statementAssign",
                  "lvalue": {
                    "id": 10,
                    "kind": "expressionId",
                    "name": {
                      "id": 9,
                      "kind": "id",
                      "name": "y",
                    },
                  },
                  "value": {
                    "id": 11,
                    "kind": "integerLiteral",
                    "value": "10",
                  },
                },
              ],
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 6 1`] = `
{
  "id": 13,
  "items": [
    {
      "id": 12,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 11,
          "kind": "handler",
          "statements": [
            {
              "id": 10,
              "kind": "statementAssign",
              "lvalue": {
                "id": 6,
                "kind": "expressionId",
                "name": {
                  "id": 5,
                  "kind": "id",
                  "name": "x",
                },
              },
              "value": {
                "id": 9,
                "kind": "expressionBinary",
                "left": {
                  "id": 7,
                  "kind": "integerLiteral",
                  "value": "5",
                },
                "op": "+",
                "right": {
                  "id": 8,
                  "kind": "integerLiteral",
                  "value": "3",
                },
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 7 1`] = `
{
  "id": 12,
  "items": [
    {
      "id": 11,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 10,
          "kind": "handler",
          "statements": [
            {
              "id": 9,
              "kind": "statementAssign",
              "lvalue": {
                "id": 6,
                "kind": "expressionId",
                "name": {
                  "id": 5,
                  "kind": "id",
                  "name": "x",
                },
              },
              "value": {
                "id": 8,
                "kind": "expressionUnary",
                "op": "-",
                "operand": {
                  "id": 7,
                  "kind": "integerLiteral",
                  "value": "5",
                },
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 8 1`] = `
{
  "id": 11,
  "items": [
    {
      "id": 10,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 9,
          "kind": "handler",
          "statements": [
            {
              "id": 8,
              "kind": "statementAssign",
              "lvalue": {
                "id": 6,
                "kind": "expressionId",
                "name": {
                  "id": 5,
                  "kind": "id",
                  "name": "x",
                },
              },
              "value": {
                "id": 7,
                "kind": "integerLiteral",
                "value": "42",
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 9 1`] = `
{
  "id": 10,
  "items": [
    {
      "id": 9,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 8,
          "kind": "handler",
          "statements": [
            {
              "expression": {
                "id": 6,
                "kind": "expressionId",
                "name": {
                  "id": 5,
                  "kind": "id",
                  "name": "myVar",
                },
              },
              "id": 7,
              "kind": "statementExpression",
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 10 1`] = `
{
  "id": 16,
  "items": [
    {
      "id": 15,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 14,
          "kind": "handler",
          "statements": [
            {
              "expression": {
                "id": 12,
                "kind": "expressionBinary",
                "left": {
                  "id": 9,
                  "kind": "expressionBinary",
                  "left": {
                    "id": 6,
                    "kind": "expressionId",
                    "name": {
                      "id": 5,
                      "kind": "id",
                      "name": "x",
                    },
                  },
                  "op": "+",
                  "right": {
                    "id": 8,
                    "kind": "expressionId",
                    "name": {
                      "id": 7,
                      "kind": "id",
                      "name": "y",
                    },
                  },
                },
                "op": "*",
                "right": {
                  "id": 11,
                  "kind": "expressionId",
                  "name": {
                    "id": 10,
                    "kind": "id",
                    "name": "z",
                  },
                },
              },
              "id": 13,
              "kind": "statementExpression",
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 11 1`] = `
{
  "id": 31,
  "items": [
    {
      "id": 30,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 29,
          "kind": "handler",
          "statements": [
            {
              "id": 10,
              "kind": "statementAssign",
              "lvalue": {
                "field": {
                  "id": 7,
                  "kind": "id",
                  "name": "x",
                },
                "id": 8,
                "kind": "expressionField",
                "struct": {
                  "id": 6,
                  "kind": "expressionId",
                  "name": {
                    "id": 5,
                    "kind": "id",
                    "name": "self",
                  },
                },
              },
              "value": {
                "id": 9,
                "kind": "integerLiteral",
                "value": "10",
              },
            },
            {
              "id": 18,
              "kind": "statementAssign",
              "lvalue": {
                "field": {
                  "id": 15,
                  "kind": "id",
                  "name": "y",
                },
                "id": 16,
                "kind": "expressionField",
                "struct": {
                  "field": {
                    "id": 13,
                    "kind": "id",
                    "name": "x",
                  },
                  "id": 14,
                  "kind": "expressionField",
                  "struct": {
                    "id": 12,
                    "kind": "expressionId",
                    "name": {
                      "id": 11,
                      "kind": "id",
                      "name": "self",
                    },
                  },
                },
              },
              "value": {
                "id": 17,
                "kind": "integerLiteral",
                "value": "10",
              },
            },
            {
              "id": 28,
              "kind": "statementAssign",
              "lvalue": {
                "field": {
                  "id": 25,
                  "kind": "id",
                  "name": "z",
                },
                "id": 26,
                "kind": "expressionField",
                "struct": {
                  "field": {
                    "id": 23,
                    "kind": "id",
                    "name": "y",
                  },
                  "id": 24,
                  "kind": "expressionField",
                  "struct": {
                    "field": {
                      "id": 21,
                      "kind": "id",
                      "name": "x",
                    },
                    "id": 22,
                    "kind": "expressionField",
                    "struct": {
                      "id": 20,
                      "kind": "expressionId",
                      "name": {
                        "id": 19,
                        "kind": "id",
                        "name": "self",
                      },
                    },
                  },
                },
              },
              "value": {
                "id": 27,
                "kind": "integerLiteral",
                "value": "10",
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 12 1`] = `
{
  "id": 13,
  "items": [
    {
      "id": 12,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 11,
          "kind": "handler",
          "statements": [
            {
              "id": 10,
              "kind": "statementAssign",
              "lvalue": {
                "field": {
                  "id": 7,
                  "kind": "id",
                  "name": "x",
                },
                "id": 8,
                "kind": "expressionField",
                "struct": {
                  "id": 6,
                  "kind": "expressionId",
                  "name": {
                    "id": 5,
                    "kind": "id",
                    "name": "global",
                  },
                },
              },
              "value": {
                "id": 9,
                "kind": "integerLiteral",
                "value": "123",
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;

exports[`Grammar should parse 13 1`] = `
{
  "id": 25,
  "items": [
    {
      "id": 24,
      "items": [
        {
          "event": {
            "id": 4,
            "kind": "parameter",
            "name": {
              "id": 2,
              "kind": "id",
              "name": "event",
            },
            "type": {
              "id": 3,
              "kind": "id",
              "name": "BlockBreak",
            },
          },
          "id": 23,
          "kind": "handler",
          "statements": [
            {
              "id": 8,
              "kind": "statementLet",
              "name": {
                "id": 5,
                "kind": "id",
                "name": "x",
              },
              "type": {
                "id": 6,
                "kind": "id",
                "name": "bool",
              },
              "value": {
                "id": 7,
                "kind": "booleanLiteral",
                "value": true,
              },
            },
            {
              "id": 12,
              "kind": "statementLet",
              "name": {
                "id": 9,
                "kind": "id",
                "name": "y",
              },
              "type": {
                "id": 10,
                "kind": "id",
                "name": "bool",
              },
              "value": {
                "id": 11,
                "kind": "booleanLiteral",
                "value": false,
              },
            },
            {
              "id": 22,
              "kind": "statementLet",
              "name": {
                "id": 13,
                "kind": "id",
                "name": "z",
              },
              "type": {
                "id": 14,
                "kind": "id",
                "name": "bool",
              },
              "value": {
                "id": 21,
                "kind": "expressionBinary",
                "left": {
                  "id": 16,
                  "kind": "expressionId",
                  "name": {
                    "id": 15,
                    "kind": "id",
                    "name": "x",
                  },
                },
                "op": "&&",
                "right": {
                  "id": 20,
                  "kind": "expressionBinary",
                  "left": {
                    "id": 18,
                    "kind": "expressionId",
                    "name": {
                      "id": 17,
                      "kind": "id",
                      "name": "y",
                    },
                  },
                  "op": "&&",
                  "right": {
                    "id": 19,
                    "kind": "booleanLiteral",
                    "value": true,
                  },
                },
              },
            },
          ],
        },
      ],
      "kind": "house",
      "name": {
        "id": 1,
        "kind": "id",
        "name": "MyHouse",
      },
    },
  ],
  "kind": "module",
}
`;
